### DISCUSSION TITLE
Is it possible that we can use JSX as a template engine?

### DISCUSSION THREAD

**Discussion by mybardaklar** [webdiscus/html-bundler-webpack-plugin#87] [Q&A]
I want to use JSX as template engine. Is it possible? If it is, how to do?

Thank you! ❤

**Comment by webdiscus** [comment]
Hello @mybardaklar,

`JSX` or `TSX` must be compiled with React.
The `*.jsx`/`*.tsx` files can be loaded in an `HTML` template via `script` tag.

For example, there is your `JSX`/`TSX` file _index.tsx_:

```tsx
import React from 'react';
import ReactDOM from 'react-dom';

const App = () => {
  return (
    <>
      <h1>Hello World!</h1>
    </>
  );
}

ReactDOM.render(<App />, document.getElementById('root'));
```

Then add this `index.tsx` file into the HTML template file `index.html`:

```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>The React App</title>
    <script src="./index.tsx" defer="defer"></script> <!-- HERE define your main JSX/TSX file -->
    <link href="./styles.scss" rel="stylesheet" /> <!-- here you can load source style files -->
  </head>
  <body>
    <div id="root"></div> <!-- HERE will be rendered your JSX file -->
  </body>
</html>
```

The webpack config:

```js
const path = require('path');
const HtmlBundlerPlugin = require('html-bundler-webpack-plugin');

module.exports = {
  mode: 'production',
  output: {
    path: path.join(__dirname, 'dist'),
  },
  resolve: {
    extensions: ['.tsx', '.ts', '.jsx', '.js'], // <= add resolving for the extensions of JSX/TSX files
  },
  plugins: [
    new HtmlBundlerPlugin({
      // define HTML templates here
      entry: {
        index: 'src/index.html', // => dist/index.html
      },
    }),
  ],
  module: {
    rules: [
      {
        test: /\.(css|scss)$/,
        use: ['css-loader', 'sass-loader'],
      },
      {
        test: /\.tsx?$/, // <= define here the rule for tsx files
        use: 'ts-loader',
        exclude: /node_modules/,
      },
    ],
  },
};
```

See please the working example [How to use JSX/TSX files with the Bundler Plugin](https://stackblitz.com/edit/webpack-react18-ts-scss?file=webpack.config.js).

**Reply by mybardaklar** [reply]
I know this but I want to use JSX as a template engine. Not SPA.

**Reply by webdiscus** [reply]
>  I want to use JSX as a template engine

How do you imagine this?
JSX file requires React and will be compiled with React.

You can create SPA with React, but `React != SPA` ;-)

You can create multiple pages website, no problem, just for each page create own HTML wrap-template:

`home.htm`
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>Homepage</title>
    <script src="./home.jsx" defer="defer"></script> <!-- HERE define your main JSX/TSX file -->
  </head>
  <body>
    <div id="root"></div> <!-- HERE will be rendered your JSX file -->
  </body>
</html>
```

`about.htm`
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <title>About</title>
    <script src="./about.jsx" defer="defer"></script> <!-- HERE define your main JSX/TSX file -->
  </head>
  <body>
    <div id="root"></div> <!-- HERE will be rendered your JSX file -->
  </body>
</html>
```

Plugin config for multiple pages:
```js
   new HtmlBundlerPlugin({
      // define HTML templates here
      entry: {
        index: 'src/home.html', // => dist/index.html
        about: 'src/about.html', // => dist/about.html
        // ...
      },
    }),
```

I can't understand what you need. Can you demonstrate your imaging?

**Reply by mybardaklar** [reply]
I do not want to create SPA and I dont want to use HTML. I want to use JSX like Astro. Astro creates MPA. JSX will built my HTML files like Pug.

**Reply by webdiscus** [reply]
Astro supported page (entrypoint) files:
[.astro](https://docs.astro.build/en/basics/astro-pages/#astro-pages) <= This is not JSX !
[.md](https://docs.astro.build/en/basics/astro-pages/#markdownmdx-pages)
.mdx (with the [MDX Integration installed](https://docs.astro.build/en/guides/integrations-guide/mdx/#installation))
[.html](https://docs.astro.build/en/basics/astro-pages/#html-pages)
.js/.ts (as [endpoints](https://docs.astro.build/en/guides/endpoints/))

Astro works same as my example above. An page is `. astro` or `.html` where can be loaded JSX. But JSX can't be used as a page template.

So, using JSX as an entrypoint is impossible, sorry.

**Reply by mybardaklar** [reply]
Okay, thank you so much for your time! :)

**Reply by webdiscus** [reply]
Thank you for the idea.

I put it in backlog, perhaps in future I implement it.
I have researched this theme, it is very complex.
The JSX file must be transpiled into JS, then compiled with React runtime. It is so complex like to make new "astro" framework.
